lambda/activity/POST/index.js

```diff
@@ -108,13 +108,156 @@ async function main(event, context, lock = null) {
       // fall through and create new record for locking.
       lock = true;
     }
-    return await handleActivity(body, lock, context);
+
+    if (event?.queryStringParameters?.isForced  == 'true') {
+      return await handleActivity(body, lock, context);
+    } else {
+      return await handleVarianceTrigger(body, lock, context);
+    }
   } catch (err) {
     logger.error(err);
     return sendResponse(400, { msg: "Invalid request" }, context);
   }
 }
 
+async function handleVarianceTrigger(body, lock, context) {
+  // Check for variances.  If true, create a variance trigger, and if not, checks
+  // if there is an existing trigger and deletes it before returning.
+  try {
+    await checkVarianceTrigger(body);
+  } catch (e) {
+    // Something blew up checking the variance.
+    // TBD: What should we do?  For now fall-through.
+    logger.error(e);
+  }
+  logger.info("Adding activity record:", body);
+  return await handleActivity(body, lock, context);
+}
+
+async function deleteVariance(subAreaId, activity, date) {
+  const params = {
+    TableName: TABLE_NAME,
+    Key: {
+      pk: { S: `variance::${subAreaId}::${activity}` },
+      sk: { S: date }
+    }
+  };
+
+  logger.info("Deleting variance record:", params);
+
+  await dynamodb.deleteItem(params).promise();
+}
+
+async function checkVarianceTrigger(body) {
+  const subAreaId   = body.subAreaId;
+  const activity    = body.activity;
+  const date        = body.date;
+  const note        = body.note;
+  const orcs        = body.orcs;
+  const subAreaName = body.subAreaName;
+  const parkName    = body.parkName;
+
+  // Create a variance field array
+  let fields = [];
+  let varianceWasTriggered = false;
+
+  // Map through all fields we care about and check their values
+  let varianceConfig = EXPORT_VARIANCE_CONFIG[activity];
+  const fieldsToCheck = Object.keys(varianceConfig);
+
+  logger.info(`Fields to check: ${fieldsToCheck}`);
+
+  // Pull up to the last 3 years for this activity type and date.
+  let records = await getPreviousYearData(3, subAreaId, activity, date);
+
+  for (const field in fieldsToCheck) {
+    logger.info(`Checking ${fieldsToCheck[field]}`);
+    let current = body?.[fieldsToCheck[field]];
+    let first = records[0]?.[fieldsToCheck[field]];
+    let second = records[1]?.[fieldsToCheck[field]];
+    let third = records[2]?.[fieldsToCheck[field]];
+
+    // Grabs the field percentage from the object
+    logger.info(`Calculating variance ${first}, ${second}, ${third}, ${current}, ${varianceConfig[fieldsToCheck[field]]}`);
+
+    if (!current || !first) {
+      // We couldn't pull any records for this date
+      logger.info('No records found - cleaning up existing variance records.');
+      await deleteVariance(subAreaId, activity, date);
+      return;
+    }
```

Comment:
This code block appears to abort the creation of a variance record for the `subarea/activity/date` if any field (`first` or `current`) is falsy. Therefore, if a field in in the body or the most recent year to check is 0 or `undefined`, existing variances are deleted and no new variance is created even if other fields trigger a variance. I think 2 things need to be changed here:

1. Falsy checks can't be done on a field to determine whether a variance gets deleted, because the field can legally be 0 or `undefined` anyways.
2. The absence of one field shouldn't trigger the deletion of the whole variance record, nor should it abort the creation of a new variance record.

Reply:
Resolved in 7aa769b6d901c706d37ad7c2b1a79f09eb5dac9f